test/e2e_node/container_restart_backoff_test.go

```diff
@@ -0,0 +1,124 @@
+	"context"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+var _ = SIGDescribe("ContainerRestartBackoff [Slow]", func() {
+	f := framework.NewDefaultFramework("container-restart-backoff-test")
+	ginkgo.Context("when a container restart", func() {
+		ginkgo.It("should be restarted the container with an exponential backoff", func() {
+			ginkgo.By("create container")
+			name := "test-container-restart-backoff"
+			pod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: name,
+				},
+				Spec: v1.PodSpec{
+					RestartPolicy: v1.RestartPolicyAlways,
+					Containers: []v1.Container{
+						{
+							Name:  "c",
+							Image: busyboxImage,
+							Command: []string{
+								"false",
+							},
+						},
+					},
+				},
+			}
+			f.PodClient().Create(pod)
+			defer f.PodClient().Delete(context.Background(), pod.Name, metav1.DeleteOptions{})
+
+			watcher, err := f.PodClient().Watch(context.Background(), metav1.ListOptions{
+				FieldSelector: fields.OneTermEqualSelector("metadata.name", pod.Name).String(),
+			})
+			framework.ExpectNoError(err, "watch pod")
+			defer watcher.Stop()
+
+			backOffPeriod := 10 // pkg/kubelet.backOffPeriod
+			expectRestartBackoffSecond := []int{
+				0,
+				10, // Period duration of retreat backoff from pkg/kubelet.backOffPeriod
+				10,
+				10, // Initial duration of retreat backoff from pkg/kubelet.backOffPeriod
+				20,
+				40,
+				80,
+				160,
+				300, // Maximum duration of retreat backoff from pkg/kubelet.MaxContainerBackOff
+				300,
+			}
+
+			ginkgo.By("watch container restart")
+			var count int32
+			var lastAt time.Time
+			for result := range watcher.ResultChan() {
+				pod := result.Object.(*v1.Pod)
+				if pod.Name != name {
+					continue
+				}
+				if len(pod.Status.ContainerStatuses) == 0 {
+					continue
+				}
+				status := pod.Status.ContainerStatuses[0]
+				terminated := status.LastTerminationState.Terminated
+				if terminated == nil {
+					continue
+				}
+				if status.RestartCount <= count {
+					continue
+				}
+				if lastAt.IsZero() && pod.Status.StartTime != nil {
+					lastAt = pod.Status.StartTime.Time
+				}
+
+				count = status.RestartCount
+				if int(count) >= len(expectRestartBackoffSecond) {
+					break
+				}
+
+				//Get actual tolerance
+				gotIntervalSecond := int(terminated.StartedAt.Sub(lastAt) / time.Second)
+				wantIntervalSecond := expectRestartBackoffSecond[count]
+				gotTolerance := gotIntervalSecond - wantIntervalSecond
+				if gotTolerance < 0 {
+					gotTolerance = -gotTolerance
+				}
+
+				framework.Logf("container restart backoff: count:%d interval:%ds startdAt:%s", count, gotIntervalSecond, terminated.StartedAt)
+
+				// Theoretically, the maximum tolerance is backOffPeriod, but in order to avoid the flakes from increasing the tolerance
+				tolerance := int(float64(backOffPeriod) * 1.5)
```

Comment:
even a 50% tolerance on a 10 second interval still seems like it could be flaky if the kubelet reconcile loop is involved

Reply:
agree with @liggitt that this could be flaky when kubelet is running large numbers of concurrent pods.  if i read the original reported issue, it looked like there were periods between restarts with no backoff, is it better to just test that?  from a cluster user perspective, we make now outward commitments on retry and backoff periods generally, its more to preserve the health of the node and node agents than for the end-user benefit.